renku/core/management/migrations/m_0003__1_jsonld.py

```diff
@@ -15,13 +15,50 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-"""Migrations for dataset."""
+"""JSON-LD dataset migrations."""
+import itertools
 import os
 import uuid
 from pathlib import Path
 
+from renku.core.models.datasets import Dataset
 
-def migrate_dataset_schema(data):
+__all__ = [
+    'migrate_datasets_metadata',
+]
+
+
+def migrate_datasets_metadata(client):
+    """Apply all initial JSON-LD migrations to datasets."""
+    jsonld_migrations = {
+        'dctypes:Dataset': [_migrate_dataset_schema, _migrate_absolute_paths],
+        'schema:Dataset': [
+            _migrate_absolute_paths, _migrate_doi_identifier,
+            _migrate_same_as_structure
+        ],
+    }
+
+    old_metadata_paths = _dataset_pre_0_3(client)
+    new_metadata_paths = (client.path / client.renku_datasets_path).rglob(
+        client.METADATA
+    )
+
+    for file_ in itertools.chain(old_metadata_paths, new_metadata_paths):
+        dataset = Dataset.from_yaml(
```

Comment:
I'm not sure we can rely on `Dataset.from_yaml` here, because the object might have change from the state of the dataset on disk. 

I think the safest way is to load the yaml to dict, then pass the dict to the migrations, and save the dict to yaml again. Basically what `from_jsonld` does internally, but without having to go through the `Dataset` class.

Migrations, especially the initial one, should depend on the yaml itself as much as possible, instead of on the classes of code, so we can change those classes without having to worry that it might break migrations (otherwise we end up with a kind of chicken and egg problem). 

Let me know if I can help, translating all our current migrations to a form that doesn't depend on Dataset code is probably quite an annoying chore.